### PR TITLE
http: add ability to add/remove query parameters

### PR DESCRIPTION
**Pull request by derekargueta**
Commit Message: Add ability to add/remove query parameters.
Additional Description:
Risk Level: Low
Testing: Included
Docs Changes: Included
Release Notes: Included
Platform Specific Features: N/A
Fixes: #2098

Signed-off-by: Derek Argueta <darguetap@gmail.com>

### CODE REVIEW

**Comment by derekargueta** [comment]
some TODOs I'd like to get into this PR:
- [ ] support command substitution, e.g. populate query params with request header value, environment variable, metadata, etc.
- [ ] looks like we probably want the same APPEND/OVERWRITE-type options as header manipulation so there's clearly defined behavior if you add a query parameter and value(s) for it already exist

**Review by purajit — APPROVED** [review]

**Comment by wbpcode** [comment]
Rather than to implement this in our core code path (and core route API), considering we have a full featured L7 filter chain, I will prefer to implement this in a filter like `header_mutation`.

/wait-any

**Comment by derekargueta** [comment]
@wbpcode I had asked about that here as I began working on this: https://github.com/envoyproxy/envoy/issues/2098#issuecomment-2083394514

**Comment by derekargueta** [comment]
I don't disagree though, if that's the desired direction I can pivot.

**Comment by derekargueta** [comment]
@wbpcode no worries! yeah after sleeping on it another advantage that came to mind might be ability to modify query parameters before/after certain requests. I don't have a concrete use-case in mind but might be useful to add parameters after certain operations e.g. auth/rate-limiting to compose higher level functionality by controlling the order of filters.

> I think you can implement it the exist `header_mutation` filter directly

interesting thought, I was initially thinking of just a new `query_parameter_mutation` filter since, while it is technically a `:path` header I feel path is generally thought of as its own construct in a request but you're right

**Comment by mattklein123** [comment]
I don't feel that strongly about it. If you both feel it would be better to do it as a filter that is fine with me. I just figured all the rest of this stuff is already in the router but this doesn't have to be.

File: envoy/http/query_params.h
Comment (wbpcode) [inline envoy/http/query_params.h:4]: Seems this is unnecessary?

File: envoy/router/router.h
Comment (wbpcode) [inline envoy/router/router.h:1129]: ditto.

File: source/common/http/null_route_impl.h
Comment (wbpcode) [inline source/common/http/null_route_impl.h:116]: ditto.

File: source/common/router/config_impl.h
Comment (wbpcode) [inline source/common/router/config_impl.h:1548]: All these changes in this file seems are unnecessary. (The typo fix could be kept.) To make PR clean could simplify the reviewing.

File: source/extensions/filters/http/query_parameter_mutation/config.h
Comment (wbpcode) [inline source/extensions/filters/http/query_parameter_mutation/config.h:16]: Seems you can use an alias name to represent  the `envoy::extensions::filters::http::query_parameter_mutation::v3::Config`.

Like:
```
using ProtoQueryParameterMutation = envoy::extensions::filters::http::query_parameter_mutation::v3::Config;
```

File: source/extensions/filters/http/query_parameter_mutation/query_params_evaluator.cc
Comment (wbpcode) [inline source/extensions/filters/http/query_parameter_mutation/query_params_evaluator.cc:14]: Prefer `make_unique`.

File: source/extensions/filters/http/query_parameter_mutation/query_params_evaluator.cc
Comment (wbpcode) [inline source/extensions/filters/http/query_parameter_mutation/query_params_evaluator.cc:18]:
```
query_params_evaluator->query_params_to_add_.emplace_back(std::make_pair(query_param.key(), query_param.value()));
```

File: source/extensions/filters/http/query_parameter_mutation/query_params_evaluator.cc
Comment (wbpcode) [inline source/extensions/filters/http/query_parameter_mutation/query_params_evaluator.cc:45]: The order of operation should be documented in the API explicitly.

**Review by wbpcode — CHANGES_REQUESTED** [review]
Hi, thanks for your contribution. Some initial comments are added to start the work.

And here are some tips about how to add a new Envoy extension:

1. A maintainer sponsor is necessary to maintain the extension. This is a simple extension. So, I can get some time to help if you are willing.
2. Please add the extension directory to the CODEOWNERS. And add yourself, maintainer sponsors to the owner list.
3. Please add metadata of this extensions in the `source/extensions/extensions_metadata.yaml`
4. Please add the related proto API to the `api/BUILD` and `api/versioning/BUILD`.
5. Please add a filter doc to the `docs/root/configuration/http/http_filters` directory and add the file to the `docs/root/configuration/http/http_filters/http_filters.rst`.

Maybe there are some other requirements, but I cannot remember now. Anyway, the CI will tell us.

File: source/common/router/config_impl.cc
Comment (derekargueta) [reply source/common/router/config_impl.cc]: ah thanks for that, was missing the link on how to get to `mostSpecificHeaderMutationWins()`

**Comment by wbpcode** [comment]
The ci result is weird. Could you merge main and (resolve the conflict) re-push again? Thanks.

/wait

**Comment by derekargueta** [comment]
working on it - need to finagle some mocking to exercise the most-specific vs least-specific mutation order

**Comment by wbpcode** [comment]
Just for curious, basically, the queries are could be treated as part of the HTTP headers. So, I initially think we will do this in the `header_mutation` filter directly. And users could use only one filter if it want to modify the headers or query. Is there any considerations that let you to make it an independent filter?

**Comment by derekargueta** [comment]
My initial thinking was how "path" is generally treated as a special first-class entity vs headers. For example path rewrite is its own configuration option rather than doing a header rewrite on the ":path" header. Same for routing, there's path matching rather than just matching on the ":path" header. As such, I was trying out having query parameters be its own filter rather than under header mutation, but having worked through this I'm starting to feel otherwise. It'd probably be simpler to just have this be under header mutation filter.

**Comment by wbpcode** [comment]
Get it. Yeah, path is special. But path is different with :path. The path sometimes (most times?) would be the part without queries of :path.

If we do this under `header_mutation`, things would be simpler for both developers and the users. You only need to add your name to the `header_mutation`'s owner list. And we can do similar things, if we want to support cookies mutation in the future.

**Comment by derekargueta** [comment]
@wbpcode assuming CI eventually passes, I've generally updated this PR to instead be a part of the Header Mutation filter.

File: api/envoy/extensions/filters/http/header_mutation/v3/header_mutation.proto
Comment (wbpcode) [inline api/envoy/extensions/filters/http/header_mutation/v3/header_mutation.proto:61]: What would be the order of the `query_parameters_to_add` and `query_parameters_to_remove` and the `request_mutations`?

If you want make removing and appending order more flexible and controllable, then you can try to wrap the append and remove to the only one `QueryMutation` and define a `repeated QueryMutation = 3;` This is why we change the `requests_to_add` and `requests_to_remove` to the `request_mutations`.

File: test/extensions/filters/http/header_mutation/BUILD
Comment (wbpcode) [inline test/extensions/filters/http/header_mutation/BUILD:59]: Please use `envoy_extension_cc_test` directly.

**Review by wbpcode — COMMENTED** [review]
Thanks so much for you hard work, it's much simpler now. And some comments are added. Mainly to the API.

(PS: if you have any concern or things want to discuss, you can touch me in the slack directly. I will check slack every morning, but may don't check the github notification in timely.

File: api/envoy/extensions/filters/http/header_mutation/v3/header_mutation.proto
Comment (wbpcode) [inline api/envoy/extensions/filters/http/header_mutation/v3/header_mutation.proto:46]: nit: one thing about the naming, I think the `query` is enough. The `query` should no several possible meanings in this context.

Comment (derekargueta) [reply api/envoy/extensions/filters/http/header_mutation/v3/header_mutation.proto]: do you just mean this variable or are you also referring to `QueryParameterValueOption` -> `QueryValueOption`, `QueryParameterMutation` -> `QueryMutation` etc? I was opting to be extra-explicit on the config api and looser in the code, but I get what you mean.

**Comment by wbpcode** [comment]
See https://github.com/envoyproxy/envoy/pull/34550 for our latest common API for all these map-like mutations. :)

File: api/envoy/config/core/v3/base.proto
Comment (wbpcode) [inline api/envoy/config/core/v3/base.proto:311]: Please no. `bytes` is used here by design.

File: api/envoy/config/common/mutation_rules/v3/mutation_rules.proto
Comment (wbpcode) [inline api/envoy/config/common/mutation_rules/v3/mutation_rules.proto:147]: these should be removed now.

**Review by wbpcode — COMMENTED** [review]
Hah, Thanks for the update. Some comments to the API is added.

File: source/extensions/filters/http/header_mutation/query_params_evaluator.cc
Comment (wbpcode) [inline source/extensions/filters/http/header_mutation/query_params_evaluator.cc:45]: Please initialize the formatter when loading configuration to avoid do this at runtime.

File: api/envoy/extensions/filters/http/header_mutation/v3/header_mutation.proto
Comment (wbpcode) [inline api/envoy/extensions/filters/http/header_mutation/v3/header_mutation.proto:26]: What is the order between the `query_parameter_mutations` and `request_mutations`? Please document it in the comment of the API.

File: api/envoy/config/core/v3/base.proto
Comment (derekargueta) [reply api/envoy/config/core/v3/base.proto:311]: doh, left an unsent github comment. I was wondering about the use of `bytes`? IIUC this makes command substitution a no-go which is the primary use-case I have in mind (appending a query param from metadata). It also is pretty inconsistent - elsewhere I see an option for a `value` string and a `raw_value` bytes.

File: source/extensions/filters/http/header_mutation/query_params_evaluator.cc
Comment (derekargueta) [reply source/extensions/filters/http/header_mutation/query_params_evaluator.cc:45]: The formatter needs the `value`?

File: api/envoy/config/core/v3/base.proto
Comment (wbpcode) [reply api/envoy/config/core/v3/base.proto:311]: > IIUC this makes command substitution a no-go which is the primary use-case I have in mind (appending a query param from metadata).

The `bytes` is used to support non-UTF8. But I think it's also OK to use it as a command formatter for Envoy. But it may not be Okay for non-C++ xDS clients.

@htuch WDYT? should we use string value as value with command formatter and use bytes value as raw value?

File: source/extensions/filters/http/header_mutation/query_params_evaluator.cc
Comment (wbpcode) [reply source/extensions/filters/http/header_mutation/query_params_evaluator.cc:45]: Yeah, you need to parse the value when you constructing `QueryParamsEvaluator`. Creating the formatter has big overhead and my throw exception. It's not allowed to do that at key data path.

File: api/envoy/config/core/v3/base.proto
Comment (wbpcode) [reply api/envoy/config/core/v3/base.proto:311]: It's same as the `HeaderValue` message. The `value` field may be a string with substitution formatter like `%REQ(CUSTOM-KEY)%`. `bytes` should be fine for C++ xDS client like Envoy in this scenario, because `bytes` and string has no essential difference. But not sure if it's OK for non-C++ client.

cc @htuch

Comment (htuch) [reply api/envoy/config/core/v3/base.proto:311]: @markdroth

**Comment by soulxu** [comment]
@wbpcode gentle ping, also @derekargueta it seems need a merge with main also

**Comment by wbpcode** [comment]
I have to say sorry to the @derekargueta .

Now, this is blocked by the API. We may did incorrect decision in past about the API.

And I will try give a final conclusion next week early.

**Comment by wbpcode** [comment]
/wait-any

Temporary label and well update it tomorrow.

File: source/extensions/filters/http/header_mutation/query_params_evaluator.cc
Comment (wbpcode) [reply source/extensions/filters/http/header_mutation/query_params_evaluator.cc:45]: I think this is problem is still not be resolved.

We should load/create the formatter when loading configuration.

You can take the header mutations implemention as an example.

/lgtm api 

/wait

**Pull request closed without merging**
